File: Assets/src/model/enemies/Octopus.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Attacks at range 2.\nRuns away if you get too close.")]
public class Octopus : AIActor {
  public override float turnPriority => task is AttackGroundTask ? 90 : base.turnPriority;

  public Octopus(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 3;
  }

  public static bool IsInRange(Entity octopus, Entity target) {
    return Util.DiamondMagnitude(target.pos - octopus.pos) <= 2;
  }

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;
    if (IsNextTo(player)) {
      return new RunAwayTask(this, player.pos, 1, true);
    }
    if (CanTargetPlayer()) {
      if (IsInRange(this, player)) {
        return new AttackGroundTask(this, player.pos, 1);
      } else {
        var chase = new ChaseTargetTask(this, player);
        chase.maxMoves = 1;
        return chase;
      }
    } else {
      return new WaitTask(this, 1);
    }
  }

  internal override (int, int) BaseAttackDamage() => (1, 2);
}

// [Serializable]
// [ObjectInfo("octopus-stinger", description: "Can attack at range 2.")]
// internal class ItemOctopusStinger : EquippableItem, IWeapon, ITargetedAction<Body> {
//   public override EquipmentSlot slot => EquipmentSlot.Weapon;
//   public override int stacksMax => int.MaxValue;

//   public (int, int) AttackSpread => (1, 2);

//   string ITargetedAction<Body>.TargettedActionName => "Attack";

//   string ITargetedAction<Body>.TargettedActionDescription => "Choose a target.";

//   void ITargetedAction<Body>.PerformTargettedAction(Player player, Entity target) {
//     player.SetTasks(new AttackGroundTask(player, target.pos, 0));
//     // player.SetTasks(new AttackTask(player, target as Body));
//   }

//   IEnumerable<Body> ITargetedAction<Body>.Targets(Player player) {
//     return player.floor.bodies.Where(b => Octopus.IsInRange(player, b));
//   }
// }

File: web/src/model/enemies/Octopus.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { AttackGroundTask } from '../tasks/AttackGroundTask';
import { ChaseTargetTask } from '../tasks/ChaseTargetTask';
import { RunAwayTask } from '../tasks/RunAwayTask';
import { WaitTask } from '../tasks/WaitTask';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Entity } from '../Entity';

/**
 * Attacks at range 2. Runs away if you get too close.
 * Port of C# Octopus.cs.
 */
export class Octopus extends AIActor {
  get turnPriority(): number {
    return this.task?.constructor.name === 'AttackGroundTask' ? 90 : super.turnPriority;
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this.hp = this._baseMaxHp = 3;
  }

  baseAttackDamage(): [number, number] {
    return [1, 2];
  }

  /** Diamond distance (manhattan) <= 2 */
  static isInRange(octopus: Entity, target: Entity): boolean {
    return Vector2Int.manhattanDistance(target.pos, octopus.pos) <= 2;
  }

  protected getNextTask(): ActorTask {
    const player = GameModelRef.main.player;
    if (this.isNextTo(player)) {
      return new RunAwayTask(this, player.pos, 1, true);
    }
    if (this.canTargetPlayer()) {
      if (Octopus.isInRange(this, player)) {
        return new AttackGroundTask(this, player.pos, 1);
      }
      const chase = new ChaseTargetTask(this, player);
      chase.maxMoves = 1;
      return chase;
    }
    return new WaitTask(this, 1);
  }
}

entityRegistry.register('Octopus', Octopus);
